src/fixed_chunks.rs

```diff
@@ -0,0 +1,535 @@
+//! assert_eq!(c.pop(), Ok(12));
+//! assert_eq!(c.pop(), Ok(13));
+//! assert_eq!(c.pop(), Ok(14));
+//! assert_eq!(c.pop(), Ok(15));
+//! ```
+//!
+//! And finally, producing and consuming chunks:
+//!
+//! ```
+//! use rtrb::{RingBuffer, fixed_chunks::ChunkProducer, fixed_chunks::ChunkConsumer};
+//!
+//! let (mut p, mut c): (ChunkProducer<_>, ChunkConsumer<_>) =
+//!     RingBuffer::with_chunks(3).of_size(2);
+//!
+//! if let Ok(mut chunk) = p.push_chunk() {
+//!     chunk.copy_from_slice(&[10, 11]);
+//! } else {
+//!     unreachable!();
+//! }
+//! if let Ok(chunk) = c.pop_chunk() {
+//!     assert_eq!(chunk.as_ref(), [10, 11]);
+//! } else {
+//!     unreachable!();
+//! }
+//! assert_eq!(c.peek_chunk(), Err(rtrb::fixed_chunks::ChunkError::TooFewSlots(0)));
+//! ```
+
+use core::marker::PhantomData;
+use core::mem::MaybeUninit;
+use core::ops::{Deref, DerefMut};
+use core::sync::atomic::Ordering;
+
+use crate::{Consumer, Producer, RingBuffer};
+
+pub use crate::chunks::ChunkError;
+
+impl<T> RingBuffer<T> {
+    /// Creates a `RingBuffer` with the given number of fixed-size chunks.
+    ///
+    /// The call to `with_chunks(...)` has to be followed by `.of_size(...)`,
+    /// which will return a pair of producer and consumer.
+    /// Multiple combinations of return types are possible:
+    ///
+    /// * `(`[`Producer<T>`]`, `[`ChunkConsumer<T>`]`)`
+    /// * `(`[`ChunkProducer<T>`]`, `[`Consumer<T>`]`)`
+    /// * `(`[`ChunkProducer<T>`]`, `[`ChunkConsumer<T>`]`)`
+    ///
+    /// The desired combination can be selected with type annotations.
+    ///
+    /// The remaining possible combination `(`[`Producer<T>`]`, `[`Consumer<T>`]`)`
+    /// can be created with [`RingBuffer::new()`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use rtrb::{fixed_chunks::ChunkConsumer, Producer, RingBuffer};
+    ///
+    /// let (p, c): (Producer<_>, ChunkConsumer<_>) = RingBuffer::<f32>::with_chunks(7).of_size(8);
+    /// ```
+    ///
+    /// Typically, the producer and consumer are stored somewhere (probably in a `struct`).
+    /// Since the return types are known in this case, no type annotations are necessary
+    /// when calling `RingBuffer::with_chunks(...).of_size(...)`:
+    ///
+    /// ```
+    /// use rtrb::{fixed_chunks::ChunkProducer, Consumer, RingBuffer};
+    ///
+    /// struct OneThing {
+    ///     producer: ChunkProducer<f32>,
+    /// }
+    /// struct AnotherThing {
+    ///     consumer: Consumer<f32>,
+    /// }
+    ///
+    /// let (producer, consumer) = RingBuffer::with_chunks(100).of_size(64);
+    ///
+    /// let thing1 = OneThing { producer };
+    /// let thing2 = AnotherThing { consumer };
+    /// ```
+    ///
+    /// See the documentation of the [`fixed_chunks`](crate::fixed_chunks#examples) module
+    /// for more examples.
+    pub fn with_chunks(chunks: usize) -> Builder<T> {
+        Builder {
+            chunks,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+#[doc(hidden)]
```

Comment:
I wouldn't mark this as `#[doc(hidden)]`, but rather add some extra documentation to this struct.